mongo/mongo.go

```diff
@@ -0,0 +1,361 @@
+	if err != nil {
+		return nil, fmt.Errorf("failed listing values of collection %s: %w", collection, err)
+	}
+	defer cursor.Close(ctx)
+
+	if err := cursor.Err(); err != nil {
+		return nil, fmt.Errorf("failed listing values of %q: %w", collection, err)
+	}
+
+	var entries []*database.Entry
+
+	for cursor.Next(ctx) {
+		t := tuple{}
+
+		if err := cursor.Decode(&t); err != nil {
+			return nil, fmt.Errorf("failed decoding value: %w", err)
+		}
+
+		entries = append(entries, &database.Entry{
+			Bucket: collection,
+			Key:    t.Key,
+			Value:  t.Value,
+		})
+	}
+
+	if err := cursor.Err(); err != nil {
+		return nil, fmt.Errorf("failed listing values of collection %q: %w", collection, err)
+	}
+
+	return entries, nil
+}
+
+func (db *DB) del(ctx context.Context, collection, key []byte) error {
+	filter := createFilter("key", key)
+
+	mongoRes, err := db.db.Collection(string(collection)).DeleteOne(ctx, filter)
+	if err != nil {
+		return fmt.Errorf("failed deleting %s/%s: %w", collection, key, err)
+	}
+
+	if mongoRes.DeletedCount == 0 {
+		return fmt.Errorf("failed deleting %s/%s: %w", collection, key, database.ErrNotFound)
+	}
+
+	return nil
+}
+
+func (db *DB) cmpAndSwap(ctx context.Context, collection, key, target, newValue []byte) ([]byte, bool, error) {
+	v, err := db.get(ctx, collection, key)
+	if err != nil && !database.IsErrNotFound(err) {
+		return nil, false, err
+	}
+
+	if !bytes.Equal(v, target) {
+		return v, false, nil
+	}
+
+	if err := db.set(ctx, collection, key, newValue); err != nil {
+		return nil, false, err
+	}
+
+	return newValue, true, nil
+}
+
+func (db *DB) executeTransactions(ctx mongo.SessionContext, tx *database.Tx, session mongo.Session) error {
+	for _, op := range tx.Operations {
+		var err error
+		switch op.Cmd {
+		case database.CreateTable:
+			if err := db.CreateTable(op.Bucket); err != nil {
+				return abort(ctx, session, fmt.Errorf("failed creating table %s: %w", op.Bucket, err))
+			}
+		case database.DeleteTable:
+			if err := db.DeleteTable(op.Bucket); err != nil {
+				return abort(ctx, session, fmt.Errorf("failed deleting table %s: %w", op.Bucket, err))
+			}
+		case database.Get:
+			if op.Result, err = db.get(ctx, op.Bucket, op.Key); err != nil {
+				return abort(ctx, session, fmt.Errorf("failed getting %s/%s: %w", op.Bucket, op.Key, err))
+			}
+		case database.Set:
+			if err := db.set(ctx, op.Bucket, op.Key, op.Value); err != nil {
+				return abort(ctx, session, fmt.Errorf("failed setting %s/%s: %w", op.Bucket, op.Key, err))
+			}
+		case database.Delete:
+			if err := db.del(ctx, op.Bucket, op.Key); err != nil {
+				return abort(ctx, session, fmt.Errorf("failed deleting %s/%s: %w", op.Bucket, op.Key, err))
+			}
+		case database.CmpAndSwap:
+			op.Result, op.Swapped, err = db.cmpAndSwap(ctx, op.Bucket, op.Key, op.CmpValue, op.Value)
+			if err != nil {
+				return abort(ctx, session, fmt.Errorf("failed load-or-store on %s/%s: %w", op.Bucket, op.Key, err))
```

Comment:
I think this is better in terms of error message clarity. Errors will now say what operation an Update failed to followed by the reason that operation failed.